v3/src/components/graph/axis.tsx

```diff
@@ -0,0 +1,173 @@
+          .classed('dragging', true)
+      },
+
+      onDilateStart: D3Handler = (event: { x: number, y: number }) => {
+        select(this)
+          .classed('dragging', true)
+        scaleAtStart = scale?.copy()
+        lowerAtStart = scaleAtStart.domain()[0]
+        upperAtStart = scaleAtStart.domain()[1]
+        dilationAnchorCoord = Number(orientation === 'bottom' ? scale?.invert(event.x) : scale?.invert(event.y))
+        dragging = true
+      },
+
+      onLowerDilateDrag = (event: { x: number, y: number, dx: number, dy: number }) => {
+        const delta = -(orientation === 'bottom' ? event.dx : event.dy)
+        if (dragging && delta !== 0) {
+          setCounter(count => ++count)
+          const
+            x2 = orientation === 'bottom' ? scaleAtStart.invert(event.x) : scaleAtStart.invert(event.y),
+            ratio = (upperAtStart - x2) / (upperAtStart - dilationAnchorCoord),
+            newRange = (upperAtStart - lowerAtStart) / ratio,
+            newLowerBound = upperAtStart - newRange
+          scale?.domain([newLowerBound, upperAtStart])
+        }
+      },
+
+      onDragTranslate = (event: { dx: number; dy: number }) => {
+        const delta = -(orientation === 'bottom' ? event.dx : event.dy)
+        if (delta !== 0) {
+          const scaleDomain = scale?.domain() || [0, 1],
+            worldDelta = Number(scale?.invert(delta)) - Number(scale?.invert(0))
+          setCounter(count => ++count)
+          scale?.domain([scaleDomain[0] + worldDelta, scaleDomain[1] + worldDelta])
+        }
+      },
+
+      onUpperDilateDrag = (event: { x: number, y: number, dx: number, dy: number }) => {
+        setCounter(count => ++count)
+        const delta = (orientation === 'bottom' ? event.dx : event.dy)
+        if (dragging && delta !== 0) {
+          const
+            x2 = orientation === 'bottom' ? scaleAtStart.invert(event.x) : scaleAtStart.invert(event.y),
+            ratio = (x2 - lowerAtStart) / (dilationAnchorCoord - lowerAtStart),
+            newRange = (upperAtStart - lowerAtStart) / ratio,
+            newUpperBound = lowerAtStart + newRange
+          scale?.domain([lowerAtStart, newUpperBound])
+        }
+      },
+
+      onDragEnd = () => {
+        select(this)
+          .classed('dragging', false)
+        scaleAtStart = null
+        dragging = false
+      }
+
+    if (axisRef?.current) {
+      const theAxis = select(axisRef.current)
+        .attr("transform", props.axisProps.transform)
+        // @ts-expect-error null => SVGSVGElement
+        .call(axis(scale))
+
+      // Add three rects in which the user can drag to dilate or translate the scale
+      const tLength = props.axisProps.length !== undefined ? props.axisProps.length : 0,
+        classPrefix = orientation === 'bottom' ? 'h' : 'v',
+        numbering = orientation === 'bottom' ? [0, 1, 2] : [2, 1, 0],
+        classPostfixes = orientation === 'bottom' ? ['lower-dilate', 'translate', 'upper-dilate'] :
```

Comment:
When there's an axis model, it should be able to some of these distinctions internal to the model.